src/routing_message_filter.rs

```diff
@@ -38,15 +45,23 @@ impl RoutingMessageFilter {
 
         RoutingMessageFilter {
             incoming: MessageFilter::with_expiry_duration(incoming_duration),
+            incoming_route: MessageFilter::with_expiry_duration(incoming_duration),
             outgoing: LruCache::with_expiry_duration(outgoing_duration),
         }
     }
 
     // Filter incoming `RoutingMessage`. Return the number of times this specific message has been
     // seen, including this time.
     // TODO - refactor to avoid cloning `msg` as `MessageFilter` only holds the hash of the tuple.
-    pub fn filter_incoming(&mut self, msg: &RoutingMessage) -> usize {
-        self.incoming.insert(msg)
+    pub fn filter_incoming(&mut self, msg: &RoutingMessage, route: u8) -> FilteringResult {
+        let known_msg = self.incoming.insert(msg) > 1;
+        let known_msg_rt = self.incoming_route.insert(&(msg.clone(), route)) > 1;
+        match (known_msg, known_msg_rt) {
+            (false, false) => FilteringResult::NewMessage,
+            (true, false) => FilteringResult::KnownMessage,
+            (true, true) => FilteringResult::KnownMessageAndRoute,
+            _ => unreachable!(),
```

Comment:
I'm not entirely sure this is unreachable. Couldn't the message somehow just have timed out int `known_msg` before insertion, but not yet in `known_msg_rt`? Maybe this should just also return `KnownMessageAndRoute`?